text/0000-gen-fn.md

```diff
@@ -0,0 +1,785 @@
+
+Is there a possibility for implementing traits like `DoubleEndedIterator`, `ExactSizeIterator` at all?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## `yield from` (forwarding operation)
+
+Python has the ability to `yield from` an iterator.
+Effectively this is syntax sugar for looping over all elements of the iterator and yielding them individually.
+There are infinite options to choose from if we want such a feature, so I'm listing general ideas:
+
+### Do nothing, just use loops
+
+```rust
+for x in iter {
+    yield x
+}
+```
+
+### Language support
+
+We could do something like postfix `yield`:
+
+```rust
+iter.yield
+```
+
+Or we could use an entirely new keyword.
+
+### stdlib macro
+
+We could add a macro to the standard library and prelude.
+The macro would expand to a `for` loop + `yield`.
+
+```rust
+yield_all!(iter)
+```
+
+## Complete `Generator` support
+
+We already have a `Generator` trait on nightly that is more powerful than the `Iterator`
+API could possibly be:
+
+1. It uses `Pin<&mut Self>`, allowing self-references in the generator across yield points.
+2. It has arguments (`yield` returns the arguments passed to it in the subsequent invocations).
+
+Similar to the ideas around `async` closures,
+I think we could argue for `Generators` to be `gen` closures while `gen` blocks are a simpler concept that has no arguments and only captures variables.
+
+Either way, support for full `Generator`s should be discussed and implemented separately,
+as there are many more open questions around them beyond a simpler way to write `Iterator`s.
+
+## `async` interactions
+
+We could support using `await` in `gen async` blocks, similar to how we support `?` being used within `gen` blocks.
+We'd have similar limitations holding references held across `await` points as we do have with `yield` points.
+The solution space for `gen async` is large enough that I will not explore it here.
+This RFC's design is forward compatible with anything we decide on.
+
+At present it is only possible to have a `gen` block yield futures, but not `await` within it, similar to how
+you cannot write iterators that `await`, but that return futures from `next`.
+
+## Self-referential `gen` blocks
+
+We can allow `gen` blocks to hold borrows across `yield` points in the future.
+
+There are a few options forward (though this list is probably not complete):
+
+* Add a separate trait for pinned iteration that is also usable with `gen` and `for`.
+    * *Downside*: We would have very similar traits for the same thing.
+* Backward-compatibly add a way to change the argument type of `Iterator::next`.
+    * *Downside*: It's unclear whether this is possible.
+* Implement `Iterator` for `Pin<&mut G>` instead of for `G` directly (whatever `G` is here, but it could be a `gen` block).
+    * *Downside*: The thing being iterated over must now be pinned for the entire iteration, instead of for each invocation of `next`.
```

Comment:
I don't understand this point. Generally when something is pinned it has to stay pinned forever, you can't just unpin after one call to `next`,

Reply:
I assume the point being made is that in the process of calling `iter.next()` in Rust today, the iterator *effectively* gets pinned in an earlier stack frame throughout `next`'s execution. It does seem like a strange way to conceptualise this though.

Reply:
We've removed this point from the document.